File: dotnet/Allmark/Render/FootnoteListRenderer.cs
namespace Allmark.Render;

using Allmark.Types;

public static class FootnoteListRenderer
{
    public static OutputRenderer Create()
    {
        return new OutputRenderer
        {
            Name = "footnote_list",
            Render = Render,
        };
    }

    public static void Render(MarkdownNode node, RendererState state, bool? decode = true)
    {
        state.Output.Append("<section class=\"footnotes\">\n<ol>\n");
        var number = 1;
        foreach (var footnote in state.Footnotes)
        {
            var label = number++;
            var id = $"fn{label}";
            var href = $"#fnref{label}";
            state.Output.Append($"<li id=\"{id}\">");
            RenderChildren.Execute(footnote, state);
            var output = state.Output.ToString();
            if (output.EndsWith("</p>\n"))
            {
                state.Output.Length -= 5;
            }
            state.Output.Append($" <a href=\"{href}\" class=\"footnote-backref\">↩</a></p>\n</li>\n");
        }
        state.Output.Append("</ol>\n</section>");
    }
}

File: web/src/render-console/footnoteListRenderer.ts
import type MarkdownNode from "../types/MarkdownNode";
import type Renderer from "../types/Renderer";
import type RendererState from "../types/RendererState";
import ANSI from "./ansi";
import renderChildren from "./renderChildren";

const renderer: Renderer = {
	name: "footnote_list",
	render,
};
export default renderer;

function render(_node: MarkdownNode, state: RendererState): void {
	if (state.footnotes.length === 0) {
		return;
	}
	state.output += `\n${ANSI.dim}---${ANSI.reset}\n`;
	let number = 1;
	for (let node of state.footnotes) {
		let label = number++;
		state.output += `${ANSI.dim}[${label}]${ANSI.reset} `;
		let refNode = state.footnoteRefs[node.info!];
		if (refNode !== undefined) {
			renderChildren(refNode, state);
		}
		if (state.output.endsWith("\n")) {
			state.output = state.output.slice(0, -1);
		}
		if (state.output.endsWith("\r")) {
			state.output = state.output.slice(0, -1);
		}
		state.output += "\n";
	}
}
